Db hotels

anmalch opened the pull request:
Скрипты создания базы данных бронирования отелей, включая данные отелей, гостей, брони, комментарии гостей, посещаемых ими отелей, а закладки с отелями. 
Скрипты наполнения БД данными.
Скрипты выборок с условиями, вложенные выборки, представления и триггеры.

anmalch committed:
Создание БД

Создание таблиц, скрипты  создания структуры БД, скрипты наполнения и триггеры.

anmalch committed:
Скрипты выборок, представлений и триггеров.

Файл содержит скрипты выборок с условиями WHERE, сортировкой, вложенный скрипт, а также представления и триггеры для валидации имени и даты дня рождения.

anmalch committed:
ER Diagram

tigal commented on db_hotels.sql:
```diff
@@ -0,0 +1,480 @@
+  password_hash CHAR(65) DEFAULT NULL,
+  created_at DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
+  updated_at DATETIME DEFAULT NULL ON UPDATE CURRENT_TIMESTAMP,
+  PRIMARY KEY (id, hotel_id),
```
Тут нет смысла делать составной ключ, так как id никуда не ссылается. Можно сделать id просто первичным ключом

tigal commented on db_hotels.sql:
```diff
@@ -0,0 +1,480 @@
+  updated_at DATETIME DEFAULT NULL ON UPDATE CURRENT_TIMESTAMP,
+  PRIMARY KEY (id),
+  CONSTRAINT fk_hotels_room FOREIGN KEY (hotel_id) REFERENCES  hotels (id)
+  
```
Нам было бы логично создать отдельную таблицу с категориями комнат, чтобы там уже хранить информацию по стоимости, числу кроватей и так далее для каждого отеля. А в этой таблице хранить тип и номер номера:)

tigal commented on db_hotels.sql:
```diff
@@ -0,0 +1,480 @@
+  id BIGINT UNSIGNED NOT NULL AUTO_INCREMENT,
+  room_id BIGINT UNSIGNED NOT NULL,
+  guest_id BIGINT UNSIGNED NOT NULL,
+  guest_count INT(20) UNSIGNED NOT NULL,
```
INT(20) это не будет до 20 число, а гигантское совершенно число:) Рекомендую тут использовать SMALLINT

tigal commented on db_hotels.sql:
```diff
@@ -0,0 +1,480 @@
+  guest_count INT(20) UNSIGNED NOT NULL,
+  check_in DATETIME NOT NULL,
+  check_out DATETIME NOT NULL,
+  status VARCHAR(10) NOT NULL, -- active/canceled
```
Тогда можно сделать ENUM или BOOLEAN если статуса всего 2

tigal commented on db_hotels.sql:
```diff
@@ -0,0 +1,480 @@
+  CONSTRAINT fk_order_guest FOREIGN KEY (guest_id) REFERENCES  guests (id),
+  CONSTRAINT fk_order_booking FOREIGN KEY (booking_id) REFERENCES  bookings (id) 
+);
+
```
Неочень понимаю, чем эта таблица от bookings отличается? Фактически они отвечают за хранение одной и той же информации? Может их слить в одну?
И почему мы храним информацию о госте здесь? Это если забронировали на другого человека номер, так?

tigal replied:
Запуталась с этой таблицей, тут получается еще дублирование информации в виде room_id, которая есть уже в bookings, числа гостей

anmalch replied:
bookings - бронироние(резервирование) можно отменить, оплата еще не внесена, и один гость может оплачивать за другого человека(например, один член семьи снимает несколько комнат в отеле, а потом он оплачивает счет (orders) за эти комнаты)

tigal replied:
Можно переименовать в bill или check_out.
Можно убрать из этой таблицы guest_id, room_id, эта информация есть в букинге.
Также можно убрать number_of_nights, эта информация также есть в букинге.
number_of_adult и number_of_children рекомендую переместить в букинг

tigal replied:
Можно сделать здесь связь от 1:1 к букинг
То есть сделать booking_id PK, удалить id

tigal commented on db_hotels.sql:
```diff
@@ -0,0 +1,480 @@
+  PRIMARY KEY (id),
+  INDEX fk_order_order_details_idx (order_id),
+  CONSTRAINT fk_order_order_details FOREIGN KEY (order_id) REFERENCES  orders (id)
+  
```
Тогда стоимость необходимо сюда добавить?

tigal commented on db_hotels.sql:
```diff
@@ -0,0 +1,480 @@
+  INDEX fk_reviews_from_guest_idx (from_guest_id),
+  CONSTRAINT fk_reviews_to_hotel FOREIGN KEY (to_hotel_id) REFERENCES  hotels (id),
+  CONSTRAINT fk_from_guest FOREIGN KEY (from_guest_id) REFERENCES  guests (id),
+  CONSTRAINT grade CHECK (REGEXP_LIKE(grade, '^[1-5]{1}$'))
```
Тут необходимо ссылаться на бронирование, иначе пользователь сможет оставить отзыв на отель, в котором он не был

tigal commented on db_hotels.sql:
```diff
@@ -0,0 +1,480 @@
+  guest_id  BIGINT UNSIGNED NOT NULL,
+  comment VARCHAR (255) DEFAULT NULL,
+  created_at DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP, 
+  PRIMARY KEY (id),
```
Тут нам необходимо сделать составной первичный ключ по hotel_id, guest_id , иначе пользователь сможет дважды отель в закладки добавить

tigal commented on db_hotels.sql:
```diff
@@ -0,0 +1,480 @@
+    END IF;
+END //
+DELIMITER ;
+
```
По определению таблицы гостей имя и фамилия у нас ненулевые, эти триггеры излишни

tigal commented on db_hotels.sql:
```diff
@@ -0,0 +1,480 @@
+END //
+
+DELIMITER ;
+
```
А вот это правильный триггер:)

anmalch committed:
Update db_hotels.sql

anmalch committed:
Update db_hotels.sql

anmalch committed:
Update db_hotels.sql

anmalch committed:
Update db_hotels.sql

anmalch committed:
Update db_hotels.sql

anmalch committed:
Update db_hotels.sql

Добавила создание и наполнение таблицы room_details, в связи с этим изменила поля и наполнение  в rooms.

tigal commented on db_hotels.sql:
```diff
@@ -0,0 +1,517 @@
+
+UNLOCK TABLES;
+LOCK TABLES room_details WRITE;
+
```
Потеряли определение room_details

tigal commented on db_hotels.sql:
```diff
@@ -0,0 +1,517 @@
+  CONSTRAINT fk_booking_room FOREIGN KEY (room_id) REFERENCES  rooms (id),
+  CONSTRAINT fk_booking_guest FOREIGN KEY (guest_id) REFERENCES  guests (id)
+);  
+
```
Может быть добавить еще один статус как "завершено"? Для успешно завершенных бронирований

tigal commented on db_hotels.sql:
```diff
@@ -0,0 +1,517 @@
+
+DROP TABLE IF EXISTS bookmarks;
+CREATE TABLE bookmarks (
+  id BIGINT UNSIGNED NOT NULL AUTO_INCREMENT,
```
Можно id отсюда удалить, лишняя колонка

tigal commented on db_hotels.sql:
```diff
@@ -0,0 +1,517 @@
+DROP TABLE IF EXISTS reviews;
+
+CREATE TABLE reviews (
+  id BIGINT UNSIGNED NOT NULL AUTO_INCREMENT,
```
Можно удалить id, лишняя колонка

tigal replied:
Либо делать ее primary key

tigal commented on selects.sql:
```diff
@@ -0,0 +1,151 @@
+LEFT JOIN hotel_images him
+ON
+  h.id=him.hotel_id
+GROUP BY h.name, him.file_name
```
Группировка нам здесь не нужна, мы ее никак не используем

tigal commented on selects.sql:
```diff
@@ -0,0 +1,151 @@
+WHERE guest_id =10;
+
+5. Вывести комнату(id, name) с большим количеством броней.
+
```
Если бронирований, тогда bookings. Если в которой останавливались - orders

tigal commented on selects.sql:
```diff
@@ -0,0 +1,151 @@
+   RIGHT JOIN reviews r ON h.id=r.to_hotel_id
+GROUP BY name
+ORDER BY average_rating DESC;
+
```
Вот из этого запроса кстати бы отличное представление вышло

tigal commented on selects.sql:
```diff
@@ -0,0 +1,151 @@
+    END IF;
+END //
+DELIMITER ;
+
```
По определению first_name и last_name не могут быть пустыми. Удалите триггер и попробуйте добавить гостя с нулевой фамилией, также получите ошибку. Лучше триггер этот удалить и забыть про него:)
